Cabal/Distribution/Simple/GHC.hs

```diff
@@ -1121,15 +1136,34 @@ gbuildSources verbosity specVer tmpDir bm =
                             ++ display mainModName
                             ++ "' listed in 'other-modules' illegally!"
 
-             return   (cSources bnfo, [main],
-                       filter (/= mainModName) (exeModules exe))
+             return BuildSources {
+                        cSourcesFiles      = cSources bnfo,
+                        cxxSourceFiles    = cxxSources bnfo,
+                        inputSourceFiles   = [main],
+                        inputSourceModules = filter (/= mainModName) $ exeModules exe
+                    }
 
-          else return (cSources bnfo, [main], exeModules exe)
-        else return (main : cSources bnfo, [], exeModules exe)
+          else return BuildSources {
+                          cSourcesFiles      = cSources bnfo,
+                          cxxSourceFiles    = cxxSources bnfo,
+                          inputSourceFiles   = [main],
+                          inputSourceModules = exeModules exe
+                      }
+        else return BuildSources {
+                        cSourcesFiles      = main : cSources bnfo,
+                        cxxSourceFiles    = main : cxxSources bnfo,
```

Comment:
These two lines don't look right. What if the main module is C++? (Are C++ main modules even supported?) What if it's in the C++-incompatible part of C? Don't we need to be able to tell between a C++ and a C module and then only add it to the appropriate list?

Reply:
OK. So the main module shouldn't be added to `cxxSourceFiles` at all? This is definitely worth having a test for, with a main module that uses some C++-breaking construct.